### PR TITLE
Add automated refund endpoint

### PR DESCRIPTION
**Pull request by mario-galic**
_Best to review this PR in the [Unified view](https://github.com/guardian/invoicing-api/pull/1/files?diff=unified&w=1)._

https://trello.com/c/DKxB6M3L/1173-goal-service-layer-work-for-automated-refund

Contents

- `README.md` explains what to do in case of failure
- `lightweight-archictecture.md` is the Architectural Design Record 
- `Program.scala` contains the main business logic
- `Impl.scala` contains implementation details 
- `Lambda.scala` is a wrapper for AWS execution, 
- `CLI.scala` is a wrapper for standalone execution
- Deployment
  - Riff-Raff project: `support:invoicing-api`
  - Artifact bucket: `membership-dist/support/PROD/invoicing-api/invoicing-api.jar`
- Zuora API User: `invoicing-api+uat@guardian.co.uk`

Different approach
- Confirmed **sub-second warmup** of Scala AWS lambda due to small package size of 6.6 MB
- Lightwight design
  - **Favours**: Inlined logic, testing in production, vanilla language and infrastructure 
  - **Avoids**: Abstractions and indirections, mocked tests, heavy usage of libraries/frameworks
- Runtime assertions implemented via [`tap`](https://www.scala-lang.org/api/current/scala/util/ChainingOps.html) operator (see `Program.scala`)
    ```
        input
          .tap  { preconditions }
          .pipe { program } 
          .tap  { postconditions }
    ```

TODO in next iterations:

* See if we can reference AWS parameter store in cloudformation
* Which Zuora fields to use for GUID reference to tie refund+adjustments together

### CODE REVIEW

**Comment by mario-galic** [comment]
[GOAL: Service layer work for automated refund](https://trello.com/c/DKxB6M3L/1173-goal-service-layer-work-for-automated-refund)

File: README.md [hunk 1]
```diff
@@ -1,2 +1,83 @@
...
+![image](https://user-images.githubusercontent.com/13835317/80218522-30e82700-8639-11ea-8e40-206933e5f105.png)
+1. Click `adjust invoice items` 
+1. Make sure `Account Balance` from step 2 has remained unchanged. If there is a difference, then you have done something
+wrong and that amount will be re-collected on the next payment run.
+
+### How to fix a failed automated refund attempt
+
+Urgent alarm will be emailed to fulfilment.dev which needs to be actioned like so:
+
+1. First familiarise yourself with above section explaining how to manually apply refund?
```
Comment (kelvin-chappell) [inline README.md:67]:
```suggestion
1. First familiarise yourself with above section explaining how to manually apply refund
```

File: build.sbt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  .settings(
+    name := "invoicing-api",
+    description := "Zuora Invoice management for supporters (refund, etc.)",
+    version := "0.1.0",
+    organization := "com.gu",
+    organizationName := "The Guardian",
+    scalaVersion := "2.13.1",
+    libraryDependencies ++= List(
+      "org.scalaj"      %% "scalaj-http"  % "2.4.2",
+      "com.lihaoyi"     %% "upickle"      % "1.1.0",
```
Comment (kelvin-chappell) [inline build.sbt:12]: 🎉

File: build.sbt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      "org.scalameta"   %% "munit"        % "0.7.1"   % Test,
```
Comment (kelvin-chappell) [inline build.sbt:13]: 🎉

File: src/main/scala/com/gu/invoicing/refund/Impl.scala [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    totalRefundAmount: Double,
+    refundGuid: String,
+  ): List[InvoiceItemAdjustmentWrite] = {
+
+    /* Collect all item adjustments of a particular invoice item and return remaining amount that can be adjusted/refunded */
+    def availableAmount(invoiceItem: InvoiceItem): Option[Double] = {
+      netAdjustmentsByInvoiceItemId(adjustments).get(invoiceItem.Id) match {
+        case Some(netAdjustment) =>
+          val availableRefundableAmount = invoiceItem.ChargeAmount - netAdjustment
+          if (availableRefundableAmount <= 0.0) None else Some(availableRefundableAmount)
```
Comment (kelvin-chappell) [inline src/main/scala/com/gu/invoicing/refund/Impl.scala:163]:
```suggestion
          if (availableRefundableAmount <= 0) None else Some(availableRefundableAmount)
```

File: src/main/scala/com/gu/invoicing/refund/Impl.scala [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+      .method("POST")
+      .asString
+      .body
+      .pipe(read[List[AdjustmentResult]](_))
+  }
+
+  def joinInvoiceWithInvoiceItemsOnInvoiceIdKey(invoices: List[Invoice], invoiceItemsBySubscription: Map[String, List[InvoiceItem]]): List[(String, Invoice, List[InvoiceItem])] = {
+    for {
+      (invoiceId, invoice) <- invoices.map(invoice => invoice.Id -> invoice)
+      (`invoiceId`, invoiceItems) <- invoiceItemsBySubscription
```
Comment (kelvin-chappell) [inline src/main/scala/com/gu/invoicing/refund/Impl.scala:210]: Confusing

File: src/main/scala/com/gu/invoicing/refund/Impl.scala [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+      .header("Content-Type", "application/json")
+      .postData(s"""{"queryString": "select Id, InvoiceId, InvoiceItemName, SourceId, SourceType, Status, Type, Amount FROM InvoiceItemAdjustment where InvoiceId = '$invoiceId'"}""")
+      .method("POST")
+      .asString
+      .body
+      .pipe(read[InvoiceItemAdjustmentsQueryResult](_))
+      .records
+  }
+
+  def roundHalfUp(x: Double) = BigDecimal(x).setScale(5, BigDecimal.RoundingMode.HALF_UP).toDouble
```
Comment (kelvin-chappell) [inline src/main/scala/com/gu/invoicing/refund/Impl.scala:237]: Perhaps comment on what effect is intended here

File: src/main/scala/com/gu/invoicing/refund/Impl.scala (hunk 5 above)
Comment (mario-galic) [reply src/main/scala/com/gu/invoicing/refund/Impl.scala]: Indeed, thanks! These are my favourite review comments as they force me to find a simpler solution.

File: cfn.yaml [hunk 7]
```diff
@@ -0,0 +1,172 @@
+AWSTemplateFormatVersion: "2010-09-09"
+
+Description: Apply refund to Zura invoice
```
Comment (arausuy) [inline cfn.yaml:3]: zuora

File: src/main/scala/com/gu/invoicing/refund/Impl.scala [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+      .pipe(read[Account](_))
+      .metrics
+      .balance
+  }
+
+  def getInvoices(accountId: String): List[Invoice] = {
+    HttpWithLongTimeout(s"$zuoraApiHost/v1/action/query")
+      .header("Authorization", s"Bearer ${accessToken()}")
+      .header("Content-Type", "application/json")
+      .postData(s"""{"queryString": "select Id, Amount, Balance, InvoiceDate, InvoiceNumber, PaymentAmount, TargetDate, Status from Invoice where AccountId = '$accountId'"}""")
```
Comment (arausuy) [inline src/main/scala/com/gu/invoicing/refund/Impl.scala:63]: Would it be more readable if the body is in a val outside?

File: src/main/scala/com/gu/invoicing/refund/Impl.scala [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+  }
+
+  /**
+   * This is likely the most complicated part of the program. It decides which invoice items to adjust and
+   * by how much by taking into account any previous adjustments already made to corresponding items.
+   */
+  def spreadRefundAcrossItems(
+    invoiceItems: List[InvoiceItem],
+    adjustments: List[InvoiceItemAdjustment],
+    totalRefundAmount: Double,
```
Comment (arausuy) [inline src/main/scala/com/gu/invoicing/refund/Impl.scala:154]: BigDecimal is a better choice when using currencies

Comment (arausuy) [reply src/main/scala/com/gu/invoicing/refund/Impl.scala:154]:
```
val d1: Double = 2.61
val d2: Double = 0.62

println(d1 - d2) //1.9899999999999998
```

File: src/main/scala/com/gu/invoicing/refund/Program.scala [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Main business logic. Can be executed as CLI application or AWS Lambda
+ *
+ * Zuora cannot apply refund in one go. It is applied by first creating a refund object,
+ * and then spreading the total refund amount over possibly multiple invoice items by
+ * applying invoice item adjustment to corresponding items. Previous charge/credit adjustments
+ * have to be accounted for to calculate available adjustment amount remaining.
+ */
+object Program {
```
Comment (arausuy) [inline src/main/scala/com/gu/invoicing/refund/Program.scala:16]: This is beautiful to read

**Comment by arausuy** [comment]
PR is really good, the code is straight forward to read 👍

I think the use of `Double` will need to be looked at: http://www.javapractices.com/topic/TopicAction.do?Id=13

File: src/main/scala/com/gu/invoicing/refund/Impl.scala (hunk 9 above)
Comment (mario-galic) [reply src/main/scala/com/gu/invoicing/refund/Impl.scala:154]: I will address this in followup PR.

Comment (mario-galic) [reply src/main/scala/com/gu/invoicing/refund/Impl.scala:154]: https://github.com/guardian/invoicing-api/pull/2

File: src/main/scala/com/gu/invoicing/refund/Impl.scala (hunk 4 above)
Comment (mario-galic) [reply src/main/scala/com/gu/invoicing/refund/Impl.scala:163]: Addressed in https://github.com/guardian/invoicing-api/pull/2

**Pull request merged**